File: implementations/cpp/provekit/verifier/smt_emitter.cpp
// SPDX-License-Identifier: Apache-2.0

#include "smt_emitter.hpp"

#include <map>
#include <set>
#include <sstream>

namespace provekit::verifier {

namespace {

std::string smt_atomic_name(const std::string& n) {
    if (n == "\xe2\x89\xa0") return "distinct";  // ≠
    if (n == "\xe2\x89\xa4") return "<=";          // ≤
    if (n == "\xe2\x89\xa5") return ">=";          // ≥
    return n;
}

std::string smt_sort(const Json& s) {
    if (s.is_object()) {
        const std::string n = s.value("name", "");
        if (n == "Bool" || n == "Real" || n == "String" || n == "Int") return n;
        if (!n.empty()) return n;
    }
    return "Int";
}

std::string emit_term(const Json& t, std::string* err) {
    if (!t.is_object()) {
        *err = "non-object IR term";
        return "";
    }
    const std::string kind = t.value("kind", "");
    if (kind == "var") {
        const std::string nm = t.value("name", "");
        if (nm.empty()) { *err = "var: empty name"; return ""; }
        return nm;
    }
    if (kind == "const") {
        if (!t.contains("value")) { *err = "const: missing value"; return ""; }
        const auto& v = t["value"];
        if (v.is_number_integer()) return std::to_string(v.get<int64_t>());
        if (v.is_number_unsigned()) return std::to_string(v.get<uint64_t>());
        if (v.is_number_float()) {
            double d = v.get<double>();
            if (d == static_cast<double>(static_cast<int64_t>(d))) {
                return std::to_string(static_cast<int64_t>(d));
            }
            std::ostringstream os; os << d; return os.str();
        }
        if (v.is_boolean()) return v.get<bool>() ? "true" : "false";
        if (v.is_string()) return "\"" + v.get<std::string>() + "\"";
        *err = "const: unsupported value type";
        return "";
    }
    if (kind == "ctor") {
        const std::string name = t.value("name", "");
        if (!t.contains("args") || !t["args"].is_array()) return name;
        const auto& args = t["args"];
        if (args.empty()) return name;
        std::ostringstream os;
        os << "(" << name;
        for (const auto& a : args) {
            std::string s = emit_term(a, err);
            if (!err->empty()) return "";
            os << " " << s;
        }
        os << ")";
        return os.str();
    }
    *err = "emit_term: unknown kind '" + kind + "'";
    return "";
}

std::string emit_formula(const Json& f, std::string* err) {
    if (!f.is_object()) { *err = "non-object IR formula"; return ""; }
    const std::string kind = f.value("kind", "");
    if (kind == "atomic") {
        // New IR: atomic uses `name` (not `predicate`).
        const std::string nm = f.value("name", "");
        const std::string smt_n = smt_atomic_name(nm);
        if (!f.contains("args") || !f["args"].is_array()) { *err = "atomic: no args"; return ""; }
        std::ostringstream os;
        os << "(" << smt_n;
        for (const auto& a : f["args"]) {
            std::string s = emit_term(a, err);
            if (!err->empty()) return "";
            os << " " << s;
        }
        os << ")";
        return os.str();
    }
    // Connectives: unified `operands` array.
    if (kind == "and" || kind == "or" || kind == "not" || kind == "implies") {
        if (!f.contains("operands") || !f["operands"].is_array()) {
            *err = kind + ": missing operands";
            return "";
        }
        const auto& ops = f["operands"];
        const char* smt_op = kind == "implies" ? "=>" : kind.c_str();
        std::ostringstream os;
        os << "(" << smt_op;
        for (const auto& op : ops) {
            std::string s = emit_formula(op, err);
            if (!err->empty()) return "";
            os << " " << s;
        }
        os << ")";
        return os.str();
    }
    if (kind == "forall" || kind == "exists") {
        // Flat quantifier shape: {kind, name, sort, body}.
        const std::string vn = f.value("name", "");
        const std::string srt = f.contains("sort") ? smt_sort(f["sort"]) : "Int";
        if (!f.contains("body")) { *err = kind + ": missing body"; return ""; }
        std::string body = emit_formula(f["body"], err);
        if (!err->empty()) return "";
        return "(" + kind + " ((" + vn + " " + srt + ")) " + body + ")";
    }
    *err = "emit_formula: unknown kind '" + kind + "'";
    return "";
}

void collect_free_vars_term(const Json& t, std::map<std::string, std::string>& out, const std::set<std::string>& bound);

void collect_free_vars(const Json& f, std::map<std::string, std::string>& out, const std::set<std::string>& bound) {
    if (!f.is_object()) return;
    const std::string kind = f.value("kind", "");
    if (kind == "atomic") {
        if (f.contains("args") && f["args"].is_array()) {
            for (const auto& a : f["args"]) collect_free_vars_term(a, out, bound);
        }
        return;
    }
    if (kind == "and" || kind == "or" || kind == "not" || kind == "implies") {
        if (f.contains("operands") && f["operands"].is_array()) {
            for (const auto& op : f["operands"]) collect_free_vars(op, out, bound);
        }
        return;
    }
    if (kind == "forall" || kind == "exists") {
        std::set<std::string> nb = bound;
        nb.insert(f.value("name", ""));
        if (f.contains("body")) collect_free_vars(f["body"], out, nb);
        return;
    }
}

void collect_free_vars_term(const Json& t, std::map<std::string, std::string>& out, const std::set<std::string>& bound) {
    if (!t.is_object()) return;
    if (t.value("kind", "") == "var") {
        const std::string n = t.value("name", "");
        if (!bound.count(n)) {
            // VarTerm carries no sort under the new IR; default to Int when
            // we can't infer (the caller may override via context-aware
            // sort-derivation in a future commit).
            out[n] = "Int";
        }
    } else if (t.value("kind", "") == "ctor") {
        if (t.contains("args") && t["args"].is_array()) {
            for (const auto& a : t["args"]) collect_free_vars_term(a, out, bound);
        }
    }
}

}  // namespace

std::string SmtEmitter::Emit(const Json& ir_formula, std::string* err) {
    std::string body = emit_formula(ir_formula, err);
    if (!err->empty()) return "";

    std::map<std::string, std::string> free_vars;
    std::set<std::string> bound;
    collect_free_vars(ir_formula, free_vars, bound);

    std::ostringstream os;
    os << "(set-logic ALL)\n";
    for (const auto& [name, srt] : free_vars) {
        os << "(declare-const " << name << " " << srt << ")\n";
    }
    os << "(assert (not " << body << "))\n";
    os << "(check-sat)\n";
    return os.str();
}

std::string SmtEmitter::emit_term(const Json& term, std::string* err) {
    return ::provekit::verifier::emit_formula(term, err);
}

}  // namespace provekit::verifier

File: implementations/rust/provekit-verifier/src/smt_emitter.rs
// SPDX-License-Identifier: Apache-2.0
//
// Stage 5: smt_emitter. Render an obligation's IR to an SMT-LIB
// script. Mirrors .../verifier/smt_emitter.cpp.

use std::collections::{BTreeMap, BTreeSet};

use serde_json::Value as Json;

pub fn emit(ir_formula: &Json) -> Result<String, String> {
    let body = emit_formula(ir_formula)?;

    let mut free_vars: BTreeMap<String, String> = BTreeMap::new();
    let bound: BTreeSet<String> = BTreeSet::new();
    collect_free_vars(ir_formula, &mut free_vars, &bound);

    let mut out = String::new();
    out.push_str("(set-logic ALL)\n");
    for (name, srt) in &free_vars {
        out.push_str(&format!("(declare-const {name} {srt})\n"));
    }
    out.push_str(&format!("(assert (not {body}))\n"));
    out.push_str("(check-sat)\n");
    Ok(out)
}

fn emit_term(t: &Json) -> Result<String, String> {
    if !t.is_object() {
        return Err("non-object IR term".into());
    }
    let kind = t.get("kind").and_then(|v| v.as_str()).unwrap_or_default();
    match kind {
        "var" => {
            let n = t.get("name").and_then(|v| v.as_str()).unwrap_or_default();
            if n.is_empty() {
                return Err("var: empty name".into());
            }
            Ok(n.to_string())
        }
        "const" => {
            let v = t.get("value").ok_or("const: missing value")?;
            if let Some(i) = v.as_i64() {
                Ok(i.to_string())
            } else if let Some(u) = v.as_u64() {
                Ok(u.to_string())
            } else if let Some(b) = v.as_bool() {
                Ok(if b { "true".into() } else { "false".into() })
            } else if let Some(s) = v.as_str() {
                Ok(format!("\"{s}\""))
            } else if let Some(f) = v.as_f64() {
                if f == (f as i64 as f64) {
                    Ok((f as i64).to_string())
                } else {
                    Ok(f.to_string())
                }
            } else {
                Err("const: unsupported value type".into())
            }
        }
        "ctor" => {
            let name = t.get("name").and_then(|v| v.as_str()).unwrap_or_default();
            match t.get("args").and_then(|v| v.as_array()) {
                None => Ok(name.to_string()),
                Some(args) if args.is_empty() => Ok(name.to_string()),
                Some(args) => {
                    let mut s = String::from("(");
                    s.push_str(name);
                    for a in args {
                        s.push(' ');
                        s.push_str(&emit_term(a)?);
                    }
                    s.push(')');
                    Ok(s)
                }
            }
        }
        other => Err(format!("emit_term: unknown kind '{other}'")),
    }
}

fn smt_atomic_name(n: &str) -> &str {
    match n {
        "\u{2260}" => "distinct", // ≠
        "\u{2264}" => "<=",       // ≤
        "\u{2265}" => ">=",       // ≥
        other => other,
    }
}

fn smt_sort(s: &Json) -> String {
    if !s.is_object() {
        return "Int".into();
    }
    let n = s.get("name").and_then(|v| v.as_str()).unwrap_or_default();
    match n {
        "Bool" | "Real" | "String" | "Int" => n.to_string(),
        "" => "Int".into(),
        other => other.to_string(),
    }
}

fn emit_formula(f: &Json) -> Result<String, String> {
    if !f.is_object() {
        return Err("non-object IR formula".into());
    }
    let kind = f.get("kind").and_then(|v| v.as_str()).unwrap_or_default();
    match kind {
        "atomic" => {
            let nm = f.get("name").and_then(|v| v.as_str()).unwrap_or_default();
            let smt_n = smt_atomic_name(nm);
            let args = f
                .get("args")
                .and_then(|v| v.as_array())
                .ok_or("atomic: no args")?;
            let mut s = String::from("(");
            s.push_str(smt_n);
            for a in args {
                s.push(' ');
                s.push_str(&emit_term(a)?);
            }
            s.push(')');
            Ok(s)
        }
        "and" | "or" | "not" | "implies" => {
            let ops = f
                .get("operands")
                .and_then(|v| v.as_array())
                .ok_or_else(|| format!("{kind}: missing operands"))?;
            let smt_op = if kind == "implies" { "=>" } else { kind };
            let mut s = String::from("(");
            s.push_str(smt_op);
            for op in ops {
                s.push(' ');
                s.push_str(&emit_formula(op)?);
            }
            s.push(')');
            Ok(s)
        }
        "forall" | "exists" => {
            let vn = f.get("name").and_then(|v| v.as_str()).unwrap_or_default();
            let srt = f
                .get("sort")
                .map(smt_sort)
                .unwrap_or_else(|| "Int".into());
            let body = f
                .get("body")
                .ok_or_else(|| format!("{kind}: missing body"))?;
            let body_s = emit_formula(body)?;
            Ok(format!("({kind} (({vn} {srt})) {body_s})"))
        }
        other => Err(format!("emit_formula: unknown kind '{other}'")),
    }
}

fn collect_free_vars(
    f: &Json,
    out: &mut BTreeMap<String, String>,
    bound: &BTreeSet<String>,
) {
    if !f.is_object() {
        return;
    }
    let kind = f.get("kind").and_then(|v| v.as_str()).unwrap_or_default();
    match kind {
        "atomic" => {
            if let Some(args) = f.get("args").and_then(|v| v.as_array()) {
                for a in args {
                    collect_free_vars_term(a, out, bound);
                }
            }
        }
        "and" | "or" | "not" | "implies" => {
            if let Some(ops) = f.get("operands").and_then(|v| v.as_array()) {
                for op in ops {
                    collect_free_vars(op, out, bound);
                }
            }
        }
        "forall" | "exists" => {
            let mut nb = bound.clone();
            if let Some(n) = f.get("name").and_then(|v| v.as_str()) {
                nb.insert(n.to_string());
            }
            if let Some(b) = f.get("body") {
                collect_free_vars(b, out, &nb);
            }
        }
        _ => {}
    }
}

fn collect_free_vars_term(
    t: &Json,
    out: &mut BTreeMap<String, String>,
    bound: &BTreeSet<String>,
) {
    if !t.is_object() {
        return;
    }
    let kind = t.get("kind").and_then(|v| v.as_str()).unwrap_or_default();
    if kind == "var" {
        if let Some(n) = t.get("name").and_then(|v| v.as_str()) {
            if !bound.contains(n) {
                // VarTerm carries no sort under the new IR; default to
                // Int (mirrors the C++ peer's choice for v1).
                out.insert(n.to_string(), "Int".into());
            }
        }
    } else if kind == "ctor" {
        if let Some(args) = t.get("args").and_then(|v| v.as_array()) {
            for a in args {
                collect_free_vars_term(a, out, bound);
            }
        }
    }
}
